### DISCUSSION TITLE
QuadSource <--> TripleSource conversion

### DISCUSSION THREAD

**Discussion by damooo** [pchampin/sophia_rs#110] [Q&A] [answered]
Hello @pchampin , Currently it is possible to convert `Graph` <--> `Dataset`, But i cannot figure out how to convert a `TripleSource` <--> `QuadSource` at the streaming level, without loading statements into a `Graph` or `DataSet` respectively. At least if there is a method to convert `ScopedRioSourceTriple` <--> `ScopedRioSourceQuad`, it seems implementable outside for `TripleSource<Triple = ScopedRioSourceTriple>` and `QuadSource<Quad = ScopedRioSourceQuad>`.

It would be much useful if we can specify triplesource_fallback_graph_iri too for such conversion.

Thanks in advance for your help

**Comment by pchampin** [comment, marked as answer]
The original idea was that the following code would do the trick:
```rust
my_quad_source.map_quads(Quad::wrap_as_triple)
```
Indeed `quad::stream::MapSource` implements `TripleSource` if the passed closure produces `Triple`s (see [API doc]).
Unfortunately, this does not work in the general case, because of some arcane lifetime issue :-/ My bad.

A workaround consist in writing your own function that *copies* the terms of the quad, e.g.:
```rust
fn q2t<Q: Quad>(q: Q) -> [BoxTerm; 3] {
    [q.s().copied(), q.p().copied(), q.o().copied()]
}
```

and then
```rust
let triple_src = quad_src.map_quads(|q| q2t(q));
```

Ultimately, I hope to get the example above working (the one using `Quad:wrap_as_triple`), but that may occur only in release 0.8.0, which will use GATs (see my [blog post](http://champin.net/blog/2021/sophia-v0.7/) about this).

[API doc]: https://docs.rs/sophia_api/0.7.0/sophia_api/quad/stream/struct.MapSource.html#impl-TripleSource

**Reply by damooo** [reply]
Thanks for that @pchampin .

I am trying for a parsing abstraction. As nested generics/assocciated-types getting too complicated for it, I am trying to zero-down to `ScopedRioSourceTriple` and `ScopedRioSourceQuad` as concrete types for `TripleStreamingMode`, `QuadStreamingMode` traits respectively as they are concrete types given by default parsers' quad-sources/triple-sources. Hence need to convert between `ScopedRioSourceTriple` <--> `ScopedRioSourceQuad` concretely. But there seems no public way to construct them. As i checked they have basic unit of `RioTermWrapper`, whose term field is private, and no other construction method. How should i handle this case?

**Comment by damooo** [comment]
Hello @pchampin, i do published a crate [rdf_dynsyn](https://docs.rs/rdf_dynsyn/latest/rdf_dynsyn/), (source: [@ github](https://github.com/manomayam/rdf_dynsyn)). This crate provides sophia-compatible and sophia-based rdf parsers/serializers, that can be instantiated against any of supported syntaxes dynamically at run time. 

request feedback, as i am quite new to rust.

**Reply by pchampin** [reply]
I had a quick look at the documentation, and that looks really interesting :+1: I ran into the same issue when writing [a simple CLI converter](https://github.com/pchampin/sophia_rs/blob/main/sophia/examples/convert.rs) as an example of program using Sophia, but your approach is of course much cleaner and scalable.

I'll look at it in more detail when I find some time.
